symfony/plugins/orangehrmAuthenticationPlugin/Service/ResetPasswordService.php

```diff
@@ -0,0 +1,334 @@
+
+use OrangeHRM\Admin\Dto\UserSearchFilterParams;
+use OrangeHRM\Admin\Traits\Service\UserServiceTrait;
+use OrangeHRM\Authentication\Dao\ResetPasswordDao;
+use OrangeHRM\Config\Config;
+use OrangeHRM\Core\Service\EmailService;
+use OrangeHRM\Core\Traits\ControllerTrait;
+use OrangeHRM\Core\Traits\LoggerTrait;
+use OrangeHRM\Core\Traits\ORM\EntityManagerHelperTrait;
+use OrangeHRM\Core\Traits\Service\DateTimeHelperTrait;
+use OrangeHRM\Core\Utility\Base64Url;
+use OrangeHRM\Entity\EmailConfiguration;
+use OrangeHRM\Entity\Employee;
+use OrangeHRM\Entity\ResetPassword;
+use OrangeHRM\Entity\User;
+use OrangeHRM\Framework\Routing\UrlGenerator;
+use OrangeHRM\Framework\Services;
+
+class ResetPasswordService
+{
+    use DateTimeHelperTrait;
+    use EntityManagerHelperTrait;
+    use LoggerTrait;
+    use ControllerTrait;
+    use UserServiceTrait;
+
+
+
+    public const RESET_PASSWORD_TOKEN_RANDOM_BYTES_LENGTH = 16;
+
+    protected ?ResetPasswordDao $resetPasswordDao = null;
+    protected ?EmailService  $emailService=null;
+
+
+    /**
+     * @return EmailService
+     */
+    public function getEmailService(): EmailService
+    {
+        if (!$this->emailService instanceof EmailService) {
+            $this->emailService = new EmailService();
+        }
+        return $this->emailService;
+    }
+
+
+    /**
+     * @return ResetPasswordDao
+     */
+    public function getResetPasswordDao(): ResetPasswordDao
+    {
+        if (!$this->resetPasswordDao instanceof ResetPasswordDao) {
+            $this->resetPasswordDao = new ResetPasswordDao();
+        }
+        return $this->resetPasswordDao;
+    }
+
+    /**
+     * @param ResetPassword $resetPassword
+     * @return float
+     */
+    public function hasPasswordResetRequestNotExpired(ResetPassword $resetPassword): float
+    {
+        $strExpireTime = strtotime($resetPassword->getResetRequestDate()->format('Y-m-d H:i:s'));
+        $strCurrentTime = strtotime(date("Y-m-d H:i:s"));
+        return floor(($strCurrentTime - $strExpireTime) / (60 * 60 * 24));
+    }
+
+    /**
+     *
+     * @param string $resetCode
+     * @return array
+     */
+    public function extractPasswordResetMetaData(string $resetCode): array
+    {
+        $code = Base64Url::decode($resetCode);
+
+        $metaData = explode('#SEPARATOR#', $code);
+
+        array_pop($metaData);
+
+        return $metaData;
+    }
+
+    /**
+     * @param string $templateFile
+     * @param array $placeholders
+     * @param array $replacements
+     * @return array|string|string[]|null
+     */
+    public function generateEmailBody(string $templateFile, array $placeholders, array $replacements)
+    {
+        $body = file_get_contents(
+            Config::get(
+                Config::PLUGINS_DIR
+            ) . '/orangehrmAuthenticationPlugin/config/data' . '//' . $templateFile
+        );
+
+        foreach ($placeholders as $key => $value) {
+            $placeholders[$key] = "/\{{$value}\}/";
+        }
```

Comment:
:thinking:

Reply:
they only used  preg_replace whether replace variable when they have curly brackets like this
{resetLink} replace as http//localhost......